datafusion/core/src/physical_optimizer/sanity_checker.rs

```diff
@@ -0,0 +1,152 @@
+use datafusion_physical_plan::ExecutionPlanProperties;
+
+#[derive(Default)]
+pub struct SanityCheckPlan {}
+
+impl SanityCheckPlan {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl PhysicalOptimizerRule for SanityCheckPlan {
+    fn optimize(
+        &self,
+        plan: Arc<dyn ExecutionPlan>,
+        config: &ConfigOptions,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        plan.transform_down(|p| {
+            check_partition_requirements(check_sort_requirements(p)?.data)
+        })
+        .data()
+    }
+
+    fn name(&self) -> &str {
+        "SanityCheckPlan"
+    }
+
+    fn schema_check(&self) -> bool {
+        true
+    }
+}
+
+pub fn check_sort_requirements(
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+    let children = plan.children();
+    let sort_reqs = plan.required_input_ordering();
+    let children_len = children.len();
+    // TODO: Use izip!.
+    for i in 0..children_len {
+        let child = &children[i];
+        let child_sort_req = &sort_reqs[i];
+
+        // No requirement for the child
+        if child_sort_req.is_none() {
+            continue;
+        }
+
+        let child_sort_req = child_sort_req.as_ref().unwrap();
+        if !child
+            .equivalence_properties()
+            .ordering_satisfy_requirement(&child_sort_req)
+        {
+            return plan_err!(
+                "Child: {:?} does not satisfy parent order requirements",
+                child
+            );
+        }
+    }
+
+    Ok(Transformed::no(plan))
+}
+
+pub fn check_partition_requirements(
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+    Ok(Transformed::no(plan))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::physical_optimizer::test_utils::{
+        bounded_window_exec, memory_exec, sort_exec, sort_expr_options
+    };
+    use arrow::compute::SortOptions;
+    use arrow::datatypes::{DataType, Field, Schema, SchemaRef};
+    use datafusion_common::Result;
+
+    fn create_test_schema() -> Result<SchemaRef> {
+        let c9_column = Field::new("c9", DataType::Int32, true);
+        let schema = Arc::new(Schema::new(vec![c9_column]));
+        Ok(schema)
+    }
+
+    #[tokio::test]
+    async fn test_bw_sort_requirement() -> Result<()> {
+        let schema = create_test_schema()?;
+        let source = memory_exec(&schema);
+        let sort_exprs = vec![sort_expr_options(
+            "c9",
+            &source.schema(),
+            SortOptions {
+                descending: false,
+                nulls_first: false,
+            },
+        )];
+        let sort = sort_exec(sort_exprs.clone(), source);
+        let bw = bounded_window_exec("c9", sort_exprs, sort);
+        let sanity_checker = SanityCheckPlan::new();
+        let opts = ConfigOptions::default();
+        assert_eq!(sanity_checker.optimize(bw, &opts).is_ok(), true);
```

Comment:
Additionally you can assert plan of the `bw` to make test more explicit. By this way plan will be more visible: Somethinkg like below:
```rust
let expected_plan = vec![
 // Fill Plan
];
let actual_plan = get_plan_string(&bw);
assert_eq(expected_plan, actual_plan)
```